### PR TITLE
[SPARK-8924] [MLLIB, DOCUMENTATION] Added @since tags to mllib.tree

### PR DESCRIPTION
**Pull request by BryanCutler**
Added @since tags to mllib.tree

### CODE REVIEW

**Comment by mengxr** [comment]
ok to test

**Comment by mengxr** [comment]
@manishamde @chouqin Could you make a pass?

**Comment by manishamde** [comment]
@mengxr Sure. I can take a look. 

How do we handle API modifications? Any change moves the tag to the newest version?

**Comment by mengxr** [comment]
Yes, only the newest version counts.

**Comment by mengxr** [comment]
@manishamde Could you make a pass?

File: mllib/src/main/scala/org/apache/spark/mllib/tree/GradientBoostedTrees.scala [hunk 1]
```diff
@@ -73,6 +74,7 @@ class GradientBoostedTrees(private val boostingStrategy: BoostingStrategy)
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/tree/GradientBoostedTrees.scala]: There is a run method in L61

File: mllib/src/main/scala/org/apache/spark/mllib/tree/configuration/Strategy.scala [hunk 2]
```diff
@@ -184,6 +201,7 @@ object Strategy {
   /**
    * Construct a default set of parameters for [[org.apache.spark.mllib.tree.DecisionTree]]
    * @param algo Algo.Classification or Algo.Regression
+   * @since 1.3.0
    */
   def defaultStategy(algo: Algo): Strategy = algo match {
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/tree/configuration/Strategy.scala]: This has been renamed to `defaultStrategy` (and rightly so). Please rebase with master.

File: mllib/src/main/scala/org/apache/spark/mllib/tree/model/DecisionTreeModel.scala [hunk 3]
```diff
@@ -103,19 +110,32 @@ class DecisionTreeModel(val topNode: Node, val algo: Algo) extends Serializable
...
+   * @since 1.3.0
+   */
   override def save(sc: SparkContext, path: String): Unit = {
     DecisionTreeModel.SaveLoadV1_0.save(sc, path, this)
   }
 
+  /**
+   * @since 1.3.0
+   */
   override protected def formatVersion: String = DecisionTreeModel.formatVersion
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/tree/model/DecisionTreeModel.scala]: Is this not protected?

File: mllib/src/main/scala/org/apache/spark/mllib/tree/model/InformationGainStats.scala [hunk 4]
```diff
@@ -38,11 +39,17 @@ class InformationGainStats(
...
+   * @since 1.0.0
+   */
   override def toString: String = {
     s"gain = $gain, impurity = $impurity, left impurity = $leftImpurity, " +
       s"right impurity = $rightImpurity"
   }
 
+  /**
+   * @since 1.2.0
+   */
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/tree/model/InformationGainStats.scala]: We need not have tags for hashCode, equals and toString. Please get rid of them here and everywhere.

File: mllib/src/main/scala/org/apache/spark/mllib/tree/model/Node.scala [hunk 5]
```diff
@@ -79,6 +84,7 @@ class Node (
    * predict value if node is not leaf
    * @param features feature value
    * @return predicted value
+   * @since 1.0.0
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/tree/model/Node.scala]: I think this should be 1.1.0

**Comment by MechCoder** [comment]
LGTM apart from those minor comments

File: mllib/src/main/scala/org/apache/spark/mllib/tree/GradientBoostedTrees.scala (hunk 1 above)
Comment (BryanCutler) [inline mllib/src/main/scala/org/apache/spark/mllib/tree/GradientBoostedTrees.scala]: I added the tag here

**Comment by BryanCutler** [comment]
Thanks for taking a look @MechCoder 
You mentioned we don't need tags for hashCode, equals and toString - I removed these and one tag for a toDebugString too.   I was also wondering if need tags for save/load?

cc @mengxr

**Comment by MechCoder** [comment]
It is good to keep tags for save / load (because it is useful to know when they were added for a particular model and all models since don't have a save / load option)

**Comment by mengxr** [comment]
@BryanCutler We don't need since tags for `hashCode`, `equals`, and `toString` because they are inherited from `Object`, and hence have the same since version as the class itself.

**Comment by BryanCutler** [comment]
Ok, that makes sense.  Tags on those are gone now.

**Comment by manishamde** [comment]
@mengxr Sorry, did not get a chance to review this so far. Will try to do it today.

File: mllib/src/main/scala/org/apache/spark/mllib/tree/impurity/Variance.scala [hunk 6]
```diff
@@ -57,6 +60,7 @@ object Variance extends Impurity {
   /**
    * Get this impurity instance.
    * This is useful for passing impurity parameters to a Strategy in Java.
+   * @since 1.0.0
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/tree/impurity/Variance.scala]: I think this should be 1.1.0

**Comment by MechCoder** [comment]
Apart from that minor comment, LGTM @mengxr 

Thanks for the PR @BryanCutler . And @manishamde I took the liberty of reviewing the PR since you were busy. Hope you wouldn't mind.

**Comment by mengxr** [comment]
Okay, I merged this into master and branch-1.5 so we can unblock SPARK-9864. @MechCoder Please fix the issue you found in your PR. Thanks @BryanCutler for adding versions and @MechCoder for review!

File: mllib/src/main/scala/org/apache/spark/mllib/tree/impurity/Variance.scala (hunk 6 above)
Comment (BryanCutler) [inline mllib/src/main/scala/org/apache/spark/mllib/tree/impurity/Variance.scala]: Yup, you're right... Sorry, didn't get a chance to fix it before the merge

**Pull request closed without merging**